### ISSUE TITLE
We should change the default for using MLIR support for COMPLEX arithmetic

### ISSUE THREAD

**Issue by psteinfeld** [llvm/llvm-project#58799]
The current default for "disable-mlir-complex" is "false".  With this setting, our results for intrinsics sin, cos, acos, sqrt, and perhaps other intrinsics do not match the defaults for other compilers such as NAG and nvfortran.

I'd like to change the default to "true", in which case our results will match.  I'm not sure what the long term solution is, but I would prefer for our default behavior to match other compilers as closely as possible.

Note that the current default was set in https://reviews.llvm.org/D135882.

@DavidTruby, @vzakhari, @klausler, I welcome your feedback and greater wisdom.

I'm happy to make this change if we all agree that it's the right thing to do.

**Comment by kiranchandramohan** [comment]
The `mathRuntimeVersion` flag is set by default to `fast`. This setting causes Intrinsic Lowering to use the `math` Dialect Operations. I believe the Complex lowering follows the same setting to use `complex` Dialect Operations. If we decide to change the setting for Complex then it would be good to do the same for Math so that the behaviour is similar.
```
llvm::cl::opt<MathRuntimeVersion> mathRuntimeVersion(
    "math-runtime", llvm::cl::desc("Select math operations' runtime behavior:"),
    llvm::cl::values(
        clEnumValN(fastVersion, "fast", "use fast runtime behavior"),
        clEnumValN(relaxedVersion, "relaxed", "use relaxed runtime behavior"),
        clEnumValN(preciseVersion, "precise", "use precise runtime behavior")),
    llvm::cl::init(fastVersion));

static llvm::cl::opt<bool>
    disableMlirComplex("disable-mlir-complex",
                       llvm::cl::desc("Use libm instead of the MLIR complex "
                                      "dialect to lower complex operations"),
                       llvm::cl::init(false));
```

**Comment by psteinfeld** [comment]
> The `mathRuntimeVersion` flag is set by default to `fast`. This setting causes Intrinsic Lowering to use the `math` Dialect Operations. I believe the Complex lowering follows the same setting to use `complex` Dialect Operations. If we decide to change the setting for Complex then it would be good to do the same for Math so that the behaviour is similar.
> 
This makes sense to me.  But again, I'm pretty ignorant of the issues surrounding intrinsic math functions.

**Comment by vzakhari** [comment]
I guess switching `mathRuntimeVersion` to `precise` default is okay.  But then we would need a way to enable `fast` or `relaxed` via a driver option.  Does it sound reasonable for the short term to do this under `-fapprox-func`?

**Comment by DavidTruby** [comment]
I think this should be implied by -Ofast/-ffast-math as well once we have those. No harm in also having a driver option for the math runtime in addition though.

**Comment by vzakhari** [comment]
`-ffast-math` is just an umbrella for a set of configuration bits.  In the front-end there is no notion of `-ffast-math` in general.

I think there is more to that.  Until we can model strict FP behavior for MLIR operations, we want to generate the opaque calls in several cases. I would expect to see calls if (https://clang.llvm.org/docs/UsersManual.html#controlling-floating-point-behavior): except_behavior == strict OR fenv_access == on OR support_math_errno == on OR etc.

**Comment by psteinfeld** [comment]
Here's a test program that shows a difference when using or not using the MLIR support for COMPLEX intrinsics:
```
program test
  complex :: source = (0.5,0)
  complex :: dest
  dest = cos(source)
  print *, dest
  print *, sign(23.0, dest%im)
end program
```
When using MLIR support for COMPLEX intrinsics, it gets the output:
```
 (.87758255,0.)
 23.
```

When not using MLIR support for COMPLEX intrinsics, it gets:
```
 (.87758255,-0.)
 -23.
```

**Comment by DavidTruby** [comment]
I have changed the behaviour for complex numbers by default to use calls to libm rather than the MLIR Complex dialect in this patch: https://reviews.llvm.org/rG3681a7dd7818aa93956fe42d37e97da6882f23ae.
The MLIR complex dialect is now _only_ used when `-fapprox-func` (and by extension `-ffast-math`) are provided.

 I did not change the default value for `-math-runtime` so as not to affect Reals in the same patch, meaning that `-fmath-runtime=fast` is still the default and therefore also uses the libm calls for complex numbers currently.
